library/azure_rm_virtualmachinescalesetextension.py

```diff
@@ -0,0 +1,269 @@
+                type='str'
+            ),
+            type=dict(
+                type='str'
+            ),
+            type_handler_version=dict(
+                type='str'
+            ),
+            auto_upgrade_minor_version=dict(
+                type='bool'
+            ),
+            settings=dict(
+                type='dict'
+            ),
+            protected_settings=dict(
+                type='dict'
+            )
+        )
+
+        self.resource_group = None
+        self.name = None
+        self.location = None
+        self.publisher = None
+        self.type = None
+        self.type_handler_version = None
+        self.auto_upgrade_minor_version = None
+        self.settings = None
+        self.protected_settings = None
+        self.state = None
+
+        self.results = dict(changed=False, state=dict())
+
+        super(AzureRMVMSSExtension, self).__init__(derived_arg_spec=self.module_arg_spec,
+                                                 supports_tags=False)
+
+    def exec_module(self, **kwargs):
+        for key in list(self.module_arg_spec.keys()):
+            setattr(self, key, kwargs[key])
+
+        resource_group = None
+        response = None
+        to_be_updated = False
+
+        resource_group = self.get_resource_group(self.resource_group)
+        if not self.location:
+            self.location = resource_group.location
+
+        if self.state == 'present':
+            response = self.get_vmssextension()
+            if not response:
+                to_be_updated = True
+            else:
+                if self.settings is not None and response.get('settings') != self.settings:
+                    response['settings'] = self.settings
+                    to_be_updated = True
+
+                if self.protected_settings is not None and response.get('protected_settings') != self.protected_settings:
+                    response['protected_settings'] = self.protected_settings
+                    to_be_updated = True
+
+            if to_be_updated:
+                if not self.check_mode:
+                    response = self.create_or_update_vmssextension()
+                self.results['changed'] = True
+        elif self.state == 'absent':
+            if not self.check_mode:
+                self.delete_vmssextension()
+            self.results['changed'] = True
+
+        if response:
+            self.results['id'] = response.get('id')
+
+        return self.results
+
+    def create_or_update_vmssextension(self):
+        self.log("Creating VMSS extension {0}".format(self.name))
+        try:
+            params = self.compute_models.VirtualMachineScaleSetExtension(
+                location=self.location,
+                publisher=self.publisher,
+                type=self.type,
+                type_handler_version=self.type_handler_version,
+                auto_upgrade_minor_version=self.auto_upgrade_minor_version,
+                settings=self.settings,
```

Comment:
if not None?

Reply:
it's ok to pass none

Reply:
what if  pass None? such as settings was set previsously. user don't want to change it so not set the argument in second run, then here None will be passed into sdk api call. what will happen? nothing changed or previous setting set to None?

Reply:
None would overwrite existing value.
I have changed it now, so it won't be overwritten (in idempotence check).